Reapply "[GVN] MemorySSA for GVN: add optional AllowMemorySSA"

antoniofrighetto opened the pull request:
Original commit: https://github.com/llvm/llvm-project/commit/eb63cd62a4a1907dbd58f12660efd8244e7d81e9.

Previously reverted due to non-negligible compile-time impact in stage1-ReleaseLTO-g scenario. The issue has been addressed separately.

antoniofrighetto commented:
I think we still need MemSSA as being used while doing some PRE as part of MergeBlockIntoPredecessor, which is coherent with the current code. Updated and dropped the assert for now. The change will be squashed into a single commit.

antoniofrighetto commented:
Compile-time impact looks on track now: https://llvm-compile-time-tracker.com/compare.php?from=9d7df23f4d6537752854d54b0c4c583512b930d0&to=66fa76366674b0fb7e7c270b452796c9c2376fab&stat=instructions:u.

alinas commented on llvm/lib/Transforms/Scalar/GVN.cpp:
```diff
@@ -3329,7 +3343,8 @@ class llvm::gvn::GVNLegacyPass : public FunctionPass {
     AU.addRequired<OptimizationRemarkEmitterWrapperPass>();
-    AU.addPreserved<MemorySSAWrapperPass>();
+    if (Impl.isMemorySSAEnabled())
+      AU.addRequired<MemorySSAWrapperPass>();
```
(legacy pass so doesn't matter much, but..) MSSA is always preserved with the above change (if getAnalysisIfAvailable is kept)

alinas replied:
I meant that this line: `AU.addPreserved<MemorySSAWrapperPass>();` should be kept here, not removed.
But the diff to add the conditional required is good, could you readd please?
```
if (Impl.isMemorySSAEnabled())
      AU.addRequired<MemorySSAWrapperPass>();
```

antoniofrighetto replied:
Oops, had accidentally dropped it, restored it now, thanks for the catch!

alinas commented on llvm/lib/Transforms/Scalar/GVN.cpp:
```diff
@@ -3293,16 +3303,18 @@ class llvm::gvn::GVNLegacyPass : public FunctionPass {
     if (skipFunction(F))
       return false;
 
-    auto *MSSAWP = getAnalysisIfAvailable<MemorySSAWrapperPass>();
```
(legacy pass, but..) Keep getAnalysisIfAvailable here; if it's not already available and isMemorySSAEnabled(), then getAnalysis.

alinas commented on llvm/lib/Transforms/Scalar/GVN.cpp:
```diff
@@ -833,6 +839,8 @@ PreservedAnalyses GVNPass::run(Function &F, FunctionAnalysisManager &AM) {
   auto &LI = AM.getResult<LoopAnalysis>(F);
   auto *MSSA = AM.getCachedResult<MemorySSAAnalysis>(F);
+  if (isMemorySSAEnabled() && !MSSA)
+    MSSA = &AM.getResult<MemorySSAAnalysis>(F);
```
Can you add the assert inside this if clause (`assert (!MemDep && "...");`)?
If MSSA is already available, it's ok to use. But if it's unavailable and we pay the cost of computing from scratch, we should be able to drop the use MemDep. This shouldn't impact the compiler tracker, but it should make the pass crash if trying to turn on before implementing the switch.

antoniofrighetto replied:
That makes sense, updated, thanks!

alinas reviewed (approved):
Thanks! Let's give this another try :-)

antoniofrighetto committed:
Reapply "[GVN] MemorySSA for GVN: add optional `AllowMemorySSA`"

Original commit: eb63cd62a4a1907dbd58f12660efd8244e7d81e9

Previously reverted due to non-negligible compile-time impact in
stage1-ReleaseLTO-g scenario. The issue has been addressed by
always reusing previously computed MemorySSA results, and request
new ones only when `isMemorySSAEnabled` is set.

Co-authored-by: Momchil Velikov <momchil.velikov@arm.com>
